src/Pate/Verification/PairGraph.hs

```diff
@@ -0,0 +1,887 @@
+
+         -- Compute exit pairs
+         traceBundle bundle $ "Discovering exit pairs from " ++ (show bPair)
+         -- TODO, manifest errors here?
+         exitPairs <- PD.discoverPairs bundle
+         traceBundle bundle $ (show (length exitPairs) ++ " pairs found!")
+
+         checkObservables asm bundle d
+
+         -- Check the totality of the discovered pairs
+         tot <- checkTotality asm bundle d exitPairs
+         case tot of
+           CasesTotal ->
+             traceBundle bundle "Totality check succeeded."
+           TotalityCheckingError msg ->
+             traceBundle bundle ("Error while checking totality! " ++ msg)
+           TotalityCounterexample (oIP,oEnd) (pIP,pEnd) ->
+             traceBundle bundle $ unwords
+               ["Found extra exit while checking totality:"
+               , showHex oIP "", PPI.ppExitCase oEnd, showHex pIP "", PPI.ppExitCase pEnd
+               ]
+
+         -- Follow all the exit pairs we found
+         foldM (followExit asm bundle bPair d) gr' (zip [0 ..] exitPairs)
+      )
+
+
+checkObservables :: forall sym arch.
+  W4.Pred sym ->
+  SimBundle sym arch ->
+  AbstractDomain sym arch ->
+  EquivM sym arch ()
+checkObservables asm bundle preD =
+  withSym $ \sym ->
+    do let oMem = PS.simMem (PS.simOutState (PPa.pOriginal (PS.simOut bundle)))
+       let pMem = PS.simMem (PS.simOutState (PPa.pPatched  (PS.simOut bundle)))
+
+       oSeq <- liftIO (MT.observableEvents sym oMem)
+       pSeq <- liftIO (MT.observableEvents sym pMem)
+
+       traceBundle bundle $ unlines
+         [ "== original event trace =="
+         , show (prettySymSequence ppEvent oSeq)
+         ] 
+
+       traceBundle bundle $ unlines
+         [ "== patched event trace =="
+         , show (prettySymSequence ppEvent pSeq)
+         ]
+
+       -- TODO! actually check the equivalance of the observables
+
+
+-- TODO move into MemTrace and do it properly
+ppEvent :: LCB.IsSymInterface sym => MT.MemEvent sym ptrW -> Doc ann
+ppEvent (MT.MemOpEvent op)   = pretty "MemOp"
+ppEvent (MT.SyscallEvent ex) = pretty "SyscallEvent" <+> W4.printSymExpr ex
+
+
+data TotalityResult
+  = CasesTotal
+  | TotalityCheckingError String
+  | TotalityCounterexample (Integer,MS.MacawBlockEndCase) (Integer,MS.MacawBlockEndCase)
+
+-- TODO? should we try to share work with the followExit/widenPostcondition calls?
+checkTotality :: forall sym arch.
+  W4.Pred sym ->
+  SimBundle sym arch ->
+  AbstractDomain sym arch ->
+  [PPa.PatchPair (PB.BlockTarget arch)] ->
+  EquivM sym arch TotalityResult
+checkTotality asm bundle preD exits =
+  withSym $ \sym ->
+    do vcfg <- asks envConfig
+       eqRel <- asks envBaseEquiv
+       stackRegion <- asks (PMC.stackRegion . envCtx)
+
+       let solver = PCfg.cfgSolver vcfg
+       let saveInteraction = PCfg.cfgSolverInteractionFile vcfg
+
+       precond <- liftIO $ do
+         eqInputs <- PE.getPredomain sym stackRegion bundle eqRel (PS.specBody preD)
```

Comment:
This is an interesting point.  With this setup, I've been thinking of the variables as essentially being "bound" to their position in the graph, and the body of the spec as being "open" rather than closed as a lambda.  I think the two views can be compatible, but maybe we shouldn't mix them up...

I don't think I'll change this for the moment, but maybe we should talk about this some more.